File: backend/app/schemas/batch_video.py
from datetime import datetime
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator

Platform = Literal["youtube", "tiktok", "instagram"]


class BatchVideoRequest(BaseModel):
    product_ids: list[int] = Field(min_length=1, max_length=20)
    platforms: list[Platform] = Field(min_length=1, max_length=3)
    variants_per_platform: int = Field(ge=1, le=10)
    duration_seconds: Literal[15] = 15
    aspect_ratio: Literal["9:16"] = "9:16"
    language: str = Field(pattern=r"^[a-z]{2,3}(?:-[A-Z]{2})?$", max_length=35)
    priority: int = Field(default=50, ge=0, le=100)
    max_concurrency: int = Field(default=3, ge=1, le=20)
    creative_angle: str | None = Field(default=None, max_length=300)
    idempotency_key: str = Field(
        min_length=8, max_length=200, pattern=r"^[A-Za-z0-9][A-Za-z0-9._:-]+$"
    )

    @field_validator("creative_angle")
    @classmethod
    def clean_angle(cls, value: str | None) -> str | None:
        if value is None:
            return None
        cleaned = value.strip()
        return cleaned or None


class BatchVideoVariantPlan(BaseModel):
    product_id: int
    platform: Platform
    variant_index: int
    brand_kit_version_id: int | None
    brand_kit_version_digest: str | None
    source_digest: str


class BatchVideoPreflightRead(BaseModel):
    ready: bool
    contract_version: str
    request_digest: str
    preflight_digest: str
    expires_at: datetime
    variant_count: int
    max_concurrency: int
    variants: list[BatchVideoVariantPlan]
    current_stage_cost: Decimal
    currency: str
    cost_scope: Literal["orchestration_only"]
    downstream_provider_cost_status: Literal["NOT_ESTIMATED"]
    provider_call_count: Literal[0]
    ffmpeg_call_count: Literal[0]


class BatchVideoCreateRequest(BatchVideoRequest):
    request_digest: str = Field(pattern=r"^[0-9a-f]{64}$")
    preflight_digest: str = Field(pattern=r"^[0-9a-f]{64}$")
    preflight_expires_at: datetime
    cost_confirmed: Literal[True]


class BatchVideoVariantRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    batch_video_job_id: int
    product_id: int
    platform: str
    variant_index: int
    duration_seconds: int
    aspect_ratio: str
    language: str
    creative_angle: str | None
    brand_kit_version_id: int | None
    brand_kit_version_digest: str | None
    source_digest: str
    idempotency_key: str
    execution_job_id: int
    status: str
    result_entity_type: str | None
    result_entity_id: int | None
    safe_error_code: str | None
    created_at: datetime
    updated_at: datetime
    completed_at: datetime | None


class BatchVideoJobRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    request_digest: str
    idempotency_key: str
    status: str
    priority: int
    variant_count: int
    max_concurrency: int
    current_stage_cost: Decimal
    currency: str
    cost_scope: str
    downstream_provider_cost_status: str
    cost_confirmed: bool
    qwen_script_call_quota: int
    qwen_script_calls_reserved: int
    frozen_constraints_json: dict[str, object]
    created_at: datetime
    updated_at: datetime
    completed_at: datetime | None


class BatchVideoCreateRead(BaseModel):
    batch: BatchVideoJobRead
    variants: list[BatchVideoVariantRead]
    reused: bool

File: frontend/src/types/batchVideo.ts
import type { ExecutionJob } from "./execution";
import type { QwenScriptPreflight } from "./videoScriptVersion";

export type BatchPlatform = "youtube" | "tiktok" | "instagram";
export type BatchVariantStatus =
  | "WAITING"
  | "RUNNING"
  | "READY_FOR_SCRIPT"
  | "PAUSED"
  | "FAILED"
  | "CANCELLED";

export interface BatchVideoRequest {
  product_ids: number[];
  platforms: BatchPlatform[];
  variants_per_platform: number;
  duration_seconds: 15;
  aspect_ratio: "9:16";
  language: string;
  priority: number;
  max_concurrency: number;
  creative_angle: string | null;
  idempotency_key: string;
}

export interface BatchVideoPreflight {
  ready: boolean;
  request_digest: string;
  preflight_digest: string;
  expires_at: string;
  variant_count: number;
  max_concurrency: number;
  current_stage_cost: string;
  currency: string;
  cost_scope: "orchestration_only";
  downstream_provider_cost_status: "NOT_ESTIMATED";
  provider_call_count: 0;
  ffmpeg_call_count: 0;
}

export interface BatchVideoJob {
  id: number;
  request_digest: string;
  idempotency_key: string;
  status: string;
  priority: number;
  variant_count: number;
  max_concurrency: number;
  current_stage_cost: string;
  currency: string;
  cost_scope: string;
  downstream_provider_cost_status: string;
  qwen_script_call_quota: number;
  qwen_script_calls_reserved: number;
}

export interface BatchVideoVariant {
  id: number;
  batch_video_job_id: number;
  product_id: number;
  platform: BatchPlatform;
  variant_index: number;
  brand_kit_version_id: number | null;
  execution_job_id: number;
  status: BatchVariantStatus;
  safe_error_code: string | null;
  active_script_version_id: number | null;
  script_version_sequence: number;
}

export interface BatchVideoCreateResult {
  batch: BatchVideoJob;
  variants: BatchVideoVariant[];
  reused: boolean;
}

export interface BatchProductOption {
  id: number;
  name: string;
  brand_kit_version_id: number | null;
}

export interface BatchQwenScriptRequest {
  product_id: number;
  variant_ids: number[];
  strategy_id: number;
  copy_matrix_id: number | null;
}

export interface BatchQwenScriptPreflight extends BatchQwenScriptRequest {
  batch_id: number;
  items: QwenScriptPreflight[];
  preflight_digest: string;
  expires_at: string;
  ready_for_execution: boolean;
  estimated_provider_calls: number;
  estimated_cost_min: string;
  estimated_cost_max: string;
  wanx_image_generation_calls: number;
  happyhorse_generation_calls: number;
  qwen_tts_generation_calls: number;
  known_downstream_cost: string;
  total_known_cost_min: string;
  total_known_cost_max: string;
  currency: string;
  cost_estimate_basis: string;
  cost_estimate_complete: false;
  unpriced_cost_components: string[];
  requires_cost_confirmation: true;
  will_auto_activate_exact_results: true;
  provider_call_count: 0;
  database_writes: 0;
}

export interface BatchQwenScriptItem {
  variant_id: number;
  platform: BatchPlatform;
  status: "QUEUED" | "RUNNING" | "READY" | "FAILED" | "SUBMIT_UNKNOWN";
  job: ExecutionJob;
  script_version_id: number | null;
  active: boolean;
  safe_error_code: string | null;
}

export interface BatchQwenScriptResult {
  batch_id: number;
  status: "RUNNING" | "READY" | "PARTIAL_FAILED" | "FAILED";
  items: BatchQwenScriptItem[];
}
